packages/http-client-csharp/generator/docs/architecture.md

```diff
@@ -0,0 +1,205 @@
+- **`InputTypes`**: Strongly-typed representations of TypeSpec constructs (models, operations, etc.)
+
+### 2. **Code Model Generation Layer**
+
+- **`CodeModelGenerator`**: Abstract base class defining the generator contract and extensibility points
+- **`ScmCodeModelGenerator`**: Concrete implementation for System.ClientModel-based generators
+- **`TypeFactory`**: Factory pattern for creating output type providers from input types
+- **`GeneratorContext`**: Provides configuration and runtime context
+
+### 3. **Output Model Layer**
+
+- **`OutputLibrary`**: Container for all generated type providers
+- **`TypeProvider`**: Abstract representation of generated types (models, enums, clients)
+- **`Providers/`**: Concrete implementations for different code constructs:
+  - `ModelProvider`: Data models with properties and serialization
+  - `EnumProvider`: Enumeration types
+  - `MethodProvider`: Client methods and operations
+  - `PropertyProvider`: Model properties with accessors
+
+### 4. **Transformation Pipeline**
+
+- **`LibraryVisitor`**: Visitor pattern for traversing and modifying the output library
+- **`LibraryRewriter`**: Advanced transformation capabilities for code modification
+- **Plugin System**: MEF-based extensibility for custom transformations
+
+### 5. **Code Generation Layer**
+
+- **`Writers/`**: Responsible for converting providers to actual C# syntax
+- **`Snippets/`**: Reusable code patterns and expressions
+- **`Expressions/`**: Type-safe representation of C# expressions
+- **`Statements/`**: Type-safe representation of C# statements
+
+### 6. **Configuration & Context**
+
+- **`Configuration`**: Centralized configuration management from JSON input
+- **`GeneratorContext`**: Runtime context and dependency injection container
+- **`SourceInputModel`**: Integration with existing custom code via Roslyn analysis
+
+## Generator Pipeline
+
+The generation process follows a well-defined pipeline:
+
+1. **Initialization**
+
+   - Load configuration from `Configuration.json`
+   - Initialize generator context and MEF composition
+   - Load TypeSpec model from `tspCodeModel.json`
+
+2. **Input Processing**
+
+   - Deserialize TypeSpec JSON into strongly-typed input model
+   - Validate and normalize input data
+   - Build dependency graphs
+
+3. **Code Model Generation**
+
+   - Transform input types to output providers using TypeFactory
+   - Apply generator-specific customizations
+   - Build method signatures and type hierarchies
+
+4. **Visitor Pipeline**
+
+   - Execute registered visitors in dependency order
+   - Apply transformations, validations, and enhancements
+   - Support for both built-in and plugin-provided visitors
+
+5. **Source Integration**
+
+   - Analyze existing custom code using Roslyn
+   - Merge custom implementations with generated code
+   - Respect customization attributes and partial classes
+
+6. **Code Generation**
+
+   - Convert providers to C# syntax trees
```

Comment:
I don't think this is accurate. We write the providers to CodeFiles. The CodeFile content is then parsed into syntax trees to allow further Roslyn processing.